src/emu/diimage.h

```diff
@@ -213,7 +213,7 @@ class device_image_interface : public device_interface
 
 	const std::string &instance_name() const { return m_instance_name; }
 	const std::string &brief_instance_name() const { return m_brief_instance_name; }
-	const std::string &cannonical_instance_name() const { return m_canonical_instance_name; }
+	const char *cannonical_instance_name() const { return dynamic_cast<const device_t *>(this)->tag(); }
```

Comment:
Why is this even here?  It's just going to cause confusion.  You can always call `some_device_interface.device().tag()` anyway - you don't need a `dynamic_cast` like this in the first place.